test/built-ins/Set/prototype/union/set-like-class-order.js

```diff
@@ -0,0 +1,75 @@
+// Copyright (C) 2023 Kevin Gibbons, Anthony Frehner. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+esid: pending
+description: Set.prototype.union calls a Set-like class's methods in order
+features: [Set-methods]
+includes: [compareArray.js]
+---*/
+
+const observedOrder = [];
+const expectedOrder = [
+  "getting size",
+  "getting has",
+  "getting keys",
+  "calling keys",
+  "getting next",
+  // first iteration, has value
+  "calling next",
+  "getting done",
+  "getting value",
+  // second iteration, has value
+  "calling next",
+  "getting done",
+  "getting value",
+  // third iteration, no value; ends
+  "calling next",
+  "getting done",
+];
+
+function observableIterator() {
+  let values = ["a", "b"];
+  let index = 0;
+  return {
+    get next() {
+      observedOrder.push("getting next");
+      return function () {
+        observedOrder.push("calling next");
+        return {
+          get done() {
+            observedOrder.push("getting done");
+            return index >= values.length;
+          },
+          get value() {
+            observedOrder.push("getting value");
+            return values[index++];
+          },
+        };
+      };
+    },
+  };
+}
+
+class MySetLike {
+  get size() {
+    observedOrder.push("getting size");
+    return 2;
```

Comment:
Please observe the `ToNumber` call (e.g. `{ valueOf() { observedOrder.push("ToNumber(size)"); return 2 } }`)